AnnotatoBackend/Sources/App/WebSocketControllers/WebSocketController.swift

```diff
@@ -0,0 +1,108 @@
+import Foundation
+import Vapor
+import AnnotatoSharedLibrary
+import FluentKit
+
+class WebSocketController {
+    /// Each user has one websocket connection.
+    private static var openConnections: [String: WebSocket] = [:]
+    private static var logger = Logger(label: "WebSocketController")
+
+    static func handleIncomingConnection(req: Request, webSocket: WebSocket) {
+        guard let userId = try? ControllersUtil.getIdFromParamsAsString(request: req) else {
+            req.logger.info("Could not get user ID for request. Closing WebSocket connection...")
+            _ = webSocket.close()
+            return
+        }
+
+        req.logger.info("Client with ID: \(userId) connected!")
+        addToOpenConnections(userId: userId, webSocket: webSocket)
+
+        webSocket.onBinary(handleBinaryData(userId: userId, db: req.db))
+        webSocket.onText(handleTextData(userId: userId, db: req.db))
+
+        webSocket.onClose.whenComplete { _ in
+            req.logger.info("Client with ID: \(userId) disconnected!")
+            Self.handleClosedConnection(userId: userId, webSocket: webSocket)
+        }
+    }
+
+    static func sendToAllAppropriateClients<T: Codable>(
+        userId: String,
+        documentId: UUID,
+        db: Database,
+        message: T
+    ) async {
+        // Send to user that sent the websocket message
+        openConnections[userId]?.send(message: message)
+
+        do {
+
+            let documentShares = try await DocumentSharesDataAccess
+                .findAllRecipientsUsingDocumentId(db: db, documentId: documentId)
+
+            // Send to all other users that share the same document
+            for (userId, ws) in openConnections {
+                guard documentShares.contains(where: { $0.recipientId == userId }) else {
+                    continue
+                }
+
+                ws.send(message: message)
+            }
```

Comment:
```suggestion
            // Send to all other users that share the same document
            let recipientIds = Set(documentShares.map { $0.recipientId })
            for (userId, ws) in openConnections where recipientIds.contains(userId) {
                ws.send(message: message)
            }
```

Should we make this more efficient?